source/System.cpp

```diff
@@ -42,6 +42,23 @@ namespace {
 	const double VOLUME = 2000.;
 	// Above this supply amount, price differences taper off:
 	const double LIMIT = 20000.;
+
+	template<class Container>
+	void LoadRandomIntervalConditions(Container &container, const DataNode &child)
+	{
+		if(container.empty())
+			return;
+		shared_ptr<ConditionSet> conditions = container.back().GetConditions();
+		for(auto & grand : child)
+			if(grand.Size() != 2 || grand.Token(0) != "to" || grand.Token(1) != "spawn" || !grand.HasChildren())
+				continue;
+			else if(!conditions)
+				conditions = make_shared<ConditionSet>(grand);
+			else
+				conditions->Load(grand);
+		if(conditions)
+			container.back().SetConditions(conditions);
```

Comment:
What's going on here?

Reply:
It updates the conditions if new ones were read in by `conditions = make_shared<ConditionSet>(grand)`

Reply:
The way you've implemented this `conditions` object seems rather confusing.

`RandomEvent<T>` now has an `std::shared_ptr<ConditionSet>`.
This will be null to begin with.

When loading, you retrieve a copy of that `std::shared_ptr`, and if it is null, you create a new ConditionSet object to be managed by this pointer, and load the conditions into that, then you pass that back into the RandomEvent and copies the new address into its internal shared pointer.
If it is not null, you call the Load method of the associated ConditionSet and then pass the pointer back into RandomEvent to be copied again?
This is assuming that there is a `to spawn` child node. If there isn't, you still get retrieve a copy of hte pointer, check if it is null, and if it isn't, passit back in to be copied, which is weird.

Why not just store a ConditionSet inside the thing and return a reference?

Though, overall, I don't think this stuff should be handled by RandomEvent at all.
I think you should create a new class that extends RandomEvent and give it the condition stuff.
Or, perhaps create a class `Conditional` that only contains the ConditionSet related stuff and then create a new class that inherits from both it and RandomEvent (I expect this'll be a very empty class).

Reply:
A ConditionSet is big and expensive, so it shouldn't be created unless it is needed. There are numerous fleets and hazards in the GameData, so the storage would add up.

Having a condition in a RandomEvent makes sense, because it is the condition under which the event is allowed to occur. I could make a ConditionalRandomEvent subclass, but then there would be no RandomEvent objects in the game.

Reply:
I have cleaned up the changes to RandomEvent to be more appropriate to an object-oriented design. The changes to System.cpp are simpler too.

Reply:
A `ConditionSet` can become big and expensive when it is not empty, but I expect the default empty `ConditionSet` to be quite small with the empty vectors (and empty strings) in them only taking up space for a size-variable that contains zero, and a pointer-variable to the data that only contains a null-pointer with no data actually allocated.

All other code where a ConditionSet is used just uses the ConditionSet as-is and uses the IsEmpty() variable on the ConditionSet to check if the set is empty. If we want to optimize for size on empty ConditionSet(s), then let's do that within ConditionSet itself, so that we get the benefits of such optimization on all locations where it gets used. Using std::optional (https://en.cppreference.com/w/cpp/utility/optional) would then also seems to be more fitting for this type of optimization than std::shared_pointer (if the optimization would be needed/desired).